src/compiler/watch.ts

```diff
@@ -101,16 +101,80 @@ namespace ts {
+            });
+        return filesInError.map((fileName: string) => {
+            const diagnosticForFileName = find(diagnostics, diagnostic =>
+                diagnostic.file !== undefined && diagnostic.file.fileName === fileName
+            );
+
+            if(diagnosticForFileName !== undefined) {
+                const { line } = getLineAndCharacterOfPosition(diagnosticForFileName.file!, diagnosticForFileName.start!);
+                return {
+                    fileName,
+                    line: line + 1,
+                };
+            }
+        });
+    }
+
     export function getWatchErrorSummaryDiagnosticMessage(errorCount: number) {
         return errorCount === 1 ?
             Diagnostics.Found_1_error_Watching_for_file_changes :
             Diagnostics.Found_0_errors_Watching_for_file_changes;
     }
 
-    export function getErrorSummaryText(errorCount: number, newLine: string) {
+    export function getErrorSummaryText(
+        errorCount: number,
+        filesInError: readonly (ReportFileInError | undefined)[],
+        newLine: string
+    ) {
         if (errorCount === 0) return "";
-        const d = createCompilerDiagnostic(errorCount === 1 ? Diagnostics.Found_1_error : Diagnostics.Found_0_errors, errorCount);
-        return `${newLine}${flattenDiagnosticMessageText(d.messageText, newLine)}${newLine}${newLine}`;
+        const nonNilFiles = filesInError.filter(fileInError => fileInError !== undefined);
+        const distinctFileNamesWithLines = nonNilFiles.map(fileInError => `${fileInError!.fileName}:${fileInError!.line}`)
+            .filter((value, index, self) => self.indexOf(value) === index);
+        const d = errorCount === 1 ?
+            createCompilerDiagnostic(
+                filesInError[0] !== undefined ?
+                    Diagnostics.Found_1_error_in_1 :
+                    Diagnostics.Found_1_error,
+                errorCount,
+                distinctFileNamesWithLines[0]) :
+            createCompilerDiagnostic(
+                distinctFileNamesWithLines.length === 0 ?
+                    Diagnostics.Found_0_errors :
+                    distinctFileNamesWithLines.length === 1 ?
+                        Diagnostics.Found_0_errors_in_1_file :
+                        Diagnostics.Found_0_errors_in_1_files,
+                errorCount,
+                distinctFileNamesWithLines.length);
+        return `${newLine}${flattenDiagnosticMessageText(d.messageText, newLine)}${newLine}${newLine}${errorCount > 1 ? createTabularErrorsDisplay(nonNilFiles) : ""}`;
+    }
+
+    function createTabularErrorsDisplay(filesInError: (ReportFileInError | undefined)[]) {
+        const BASE_LEFT_PADDING = 6;
+        let tabularData = "";
+        const distinctFiles = filesInError.filter((value, index, self) => index === self.findIndex(file => file!.fileName === value!.fileName));
+        if(distinctFiles.length === 0) return tabularData;
+
+        const headerRow = "Errors   Files\n";

```

Comment:
This probably needs to be a localized message~

Reply:
If we do that, (which I think we probably do need to do) then the numbers about length etc will need to not be hardcoded anymore. For example the Chinese version of "errors" might be 1-2 characters characters wide.

Reply:
I didn't think of that, yes this should probably be localised.

@orta 
> If we do that, (which I think we probably do need to do) then the numbers about length etc will need to not be hardcoded anymore
May I ask how you mean by "hardcoded" here? Do you mean we don't need to bother so much about left padding?

Reply:
By hardcoding, I mean `const BASE_LEFT_PADDING = 6;` only makes sense when 'errors' the string is 6 characters long for example.

Armed with google translate, perhaps this is a reasonable design answer:

```
错误  案卷
  3  build/lib/builtInExtensionsCG.ts
```

```
 错误  案卷
  12  build/lib/builtInExtensionsCG.ts
 123  extensions/html-language-features/client/src/htmlClient.ts
```


```
  错误  案卷
   12  build/lib/builtInExtensionsCG.ts
  123  extensions/html-language-features/client/src/htmlClient.ts
 1234  extensions/html-language-features/client/src/htmlEmptyTagsShared.ts
```

Where tsc pulls out the longest length of a value from either the highest number of errors in a file or the translated word for "errors", allowing people to have a `MAX_INT` amount of errors because it just moves the column to the right once it hits a particular size

```
       错误  案卷
        12  build/lib/builtInExtensionsCG.ts
       123  extensions/html-language-features/client/src/htmlClient.ts
      1234  extensions/html-language-features/client/src/htmlEmptyTagsShared.ts
 123456789  extensions/html-language-features/client/src/tagClosing.ts
```

```
    Errors  Files
        12  build/lib/builtInExtensionsCG.ts
       123  extensions/html-language-features/client/src/htmlClient.ts
      1234  extensions/html-language-features/client/src/htmlEmptyTagsShared.ts
 123456789  extensions/html-language-features/client/src/tagClosing.ts
```

Does that feel reasonable to you?

Reply:
That will work for me too. I'll work on it and see what I can come up with. Thank you for the explanation.